app-typescript/components/Radio.tsx

```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+
+interface IValue {
+    value: any;
```

Comment:
Can you make it `string` instead of `any?` Try not using `any` at all. It's mainly meant for migrations or integration of libraries without typings. You shouldn't need it in your own code. You can always ping me if you get a compiler error you don't understand :slightly_smiling_face:

Reply:
Value can be string, object or boolean, because of that I used any.

Reply:
It doesn't make sense for value to boolean, because it wouldn't work to have more than 2 options. If you have 3 and their values are `true, false, true` and the current value is `true`, how would you know which label to mark as active?

Allowing objects isn't that good either because they can be compared by value or reference which might cause confusion when using the component. Can you limit allowed `value` type to `string`?

Reply:
But sometimes, especially in settings we need to allow assigning object as value. Why is that big deal not to allow this if we need that? Also, `PrimeReact` that you used for `DatePicker` is allowing objects as values.

https://github.com/primefaces/primereact/blob/master/src/components/checkbox/Checkbox.d.ts

Reply:
Can you show me an example of how would you use object values?

It doesn't matter what `PrimeReact` uses. I used the library because it was the best I could find at displaying the component. I could care less if they have bad interfaces because I wrap their components and don't expose their interfaces to consumers of `superdesk-ui-framework`.

Reply:
If you want to assign some specific object to value, you will use object as value. Not sure what is not clear about that? It is matter, because here we are talking about what should be the best and what is best implementation. So let's check how big teams focused only on developing components did it.

Reply:
> If you want to assign some specific object to value, you will use object as value. Not sure what is not clear about that?

Can you paste a simple code snippet? There might be multiple ways to implement it and I don't know what's on your mind.

Reply:
```
let object1 = {'prop1: 'vaue1', 'prop2': 'value2'};
let object2 = {'prop1: 'vaue1', 'prop2': 'value2'};
let object3 = {'prop1: 'vaue1', 'prop2': 'value2'};

<Radio value={this.state.value1} options={[{value:object1, text:'Default label'}, {value:object2, text:'Default label'}, {value:object3, text:'Default label'}} onClick={(value) => this.setState(() => ({ value1: value }))}/>
```

Reply:
This is dummy code which doesn't make it clear how would you use it in the real world. Even in this case, it's easy to pass `object1.prop1` as a value which is string.

The problem with objects is comparison. How do you compare if objects are the same in order to determine which label should be marked as active? If you do `===` comparison, you are checking by reference and will get bugs when reference changes but values stay the same. You can try serializing the object to a string, but not all objects are serializable. By using strings as values we would be avoiding the issue entirely.

Reply:
Yes, and your value will be only vaue1 and you will miss entirely value2. 

You wanted example, I wrote you example. Previously we needed this functionality, to get object assigned as value on `sd-check`, and I'm sure we will need this in feature. Because of that I'm highly recommending to allow this options. This is not bug, if you want you can go with string, but also you will have ability to use something else (for example object).

This is feature not a bug, not sure why we are taking energy on something that is so stupid. As I sad, string is recommended, but also we need to make ability to allow objects as values.

Reply:
I think even if we would need to support objects it would be good to list those options there and not use any,
because that not doing any check. undefined will work too so it won't catch typos.
so define value as `string  | number | {}`

Reply:
btw I think we had some usecase with elasticfilters, something like:
```
filterOptions = [
    {label: 'Today', value: {range: {versioncreated: {gte: 'now/d'}}}},
    {label: 'This week', value: {range: {versioncreated: {gte: 'now-1w/d'}}}},
    {label: 'Last week', value: {range: {versioncreated: {gte: 'now-2w/d', lte: 'now-1w/d'}}}},
    ...
];
```

Reply:
Native select inputs only allow strings as values. Here's how it would look with elastic filters:

```
return (
    <select value={this.state.value}>
        {
            filterOptions.map((option) =>
                <option
                    value={option.label}
                >
                    {option.label}
                </option>)
        }
    </select>
);
```

It requires you to have an ID for a value and then you can get the object by that ID.

If you want to go with objects, I suggest using generics, because `{test: 1}` can't be assigned to `string | number | {}`, since it's not an empty object. Here's an interface with generics:

```
interface IRadioProps<T> {
    options: Array<T>;
    value: T;
    onChange(value: T): void;
    getLabel(item: T): string;
    getId(item: T): string;
}
```

I'd look similar to how `select2` [is implemented](https://github.com/superdesk/superdesk-client-core/blob/develop/scripts/core/ui/components/select2.tsx#L8)

Reply:
right, it could be `string | number | object`. we could use generic too I would keep it simple:
```ts
interface IProps<T> {
    options: Array<{
        label: string;
        value: T;
    };
    value: T;
    onChange: (value: T) => void;
}
```

Reply:
How would you check if values are equal? Compare labels? If we force labels to be unique, which makes sense for radios, we could drop `getId` method in my suggested interface and I'd be as simple. I'd prefer passing a lambda to get a label instead of mapping the array, but both are fine.

Reply:
I would compare values by reference or via [lodash.isEqual](https://lodash.com/docs/4.17.15#isEqual)